src/Session.cpp

```diff
@@ -933,18 +929,19 @@ void CSession::UpdateStream(CStream& stream)
 
 void CSession::PrepareStream(CStream* stream)
 {
-  if (m_adaptiveTree->GetTreeType() != adaptive::TreeType::HLS)
+  if (!m_adaptiveTree->IsReqPrepareStream())
     return;
 
   CRepresentation* repr = stream->m_adStream.getRepresentation();
   const EVENT_TYPE startEvent = stream->m_adStream.GetStartEvent();
 
-  // Download the manifest only at first start of the stream
-  if (startEvent == EVENT_TYPE::STREAM_START)
+  // Prepare the representation when the period change usually its not needed,
+  // because the timeline is always already updated
+  if ((!m_adaptiveTree->IsChangingPeriod() || !repr->HasSegmentTimeline()) &&
+      (startEvent == EVENT_TYPE::STREAM_START || startEvent == EVENT_TYPE::STREAM_ENABLE))
```

Comment:
just a note for `!repr->HasSegmentTimeline()` condition, this imo should not be needed, at least not for HLS where multiple periods have the exact same streams to continue. I don't know how to explain myself well i try, kodi can call `CInputStreamAdaptive::OpenStream` with a stream ID that it decides itself and we cannot predict, so in the remote case that a stream X in a certain language is not available on the next period, kodi will probably select an alternative stream language with a different ID, therefore despite IsChangingPeriod is true, we still need to prepare the stream, since should not have the timeline

Reply:
Yes this shouldn't happen with HLS since we get the master m3u8 once and once only, the variants will never change and are same across all periods, however some could possibly end earlier than others in the case of a live stream (but why would they...)
For DASH anything can happen in a new period and so I think your thinking is correct and the check is valid. Ads would be a good example. They might not carry the same language tracks as the main program content.

Reply:
also i found another use case for this,
in case the requested video stream ID for some reason goes error as in the example below, Kodi core will try to call all subsequent streams until it finds one that works
```
2024-05-22 18:20:40.425 T:24788   debug <general>: AddOnLog: inputstream.adaptive: GetStreamIds()
2024-05-22 18:20:40.425 T:24788   debug <general>: AddOnLog: inputstream.adaptive: GetStream(3001)
2024-05-22 18:20:40.426 T:24788   debug <general>: AddOnLog: inputstream.adaptive: GetStream(3002)
2024-05-22 18:20:40.427 T:24788   debug <general>: AddOnLog: inputstream.adaptive: GetStream(3003)
2024-05-22 18:20:40.427 T:24788   debug <general>: AddOnLog: inputstream.adaptive: GetStream(3004)
2024-05-22 18:20:40.428 T:24788   debug <general>: AddOnLog: inputstream.adaptive: GetStream(3005)
2024-05-22 18:20:40.428 T:24788   debug <general>: AddOnLog: inputstream.adaptive: GetStream(3006)
2024-05-22 18:20:40.429 T:24788   debug <general>: CDVDDemuxClient::RequestStream(): added/updated stream 3001 with codec_id 27
2024-05-22 18:20:40.429 T:24788   debug <general>: CDVDDemuxClient::RequestStream(): added/updated stream 3002 with codec_id 27
2024-05-22 18:20:40.430 T:24788   debug <general>: CDVDDemuxClient::RequestStream(): added/updated stream 3003 with codec_id 27
2024-05-22 18:20:40.430 T:24788   debug <general>: CDVDDemuxClient::RequestStream(): added/updated stream 3004 with codec_id 27
2024-05-22 18:20:40.430 T:24788   debug <general>: CDVDDemuxClient::RequestStream(): added/updated stream 3005 with codec_id 27
2024-05-22 18:20:40.430 T:24788   debug <general>: CDVDDemuxClient::RequestStream(): added/updated stream 3006 with codec_id 86018
2024-05-22 18:20:40.438 T:24788    info <general>: Opening stream: 3001 source: 256
2024-05-22 18:20:40.438 T:24788   debug <general>: AddOnLog: inputstream.adaptive: OpenStream(3001)
2024-05-22 18:20:40.440 T:24788   debug <general>: AddOnLog: inputstream.adaptive: GetStream(3001)
2024-05-22 18:20:40.441 T:24788   error <general>: CVideoPlayerVideo::OpenStream: Codec id 27 require extradata.
2024-05-22 18:20:40.441 T:24788 warning <general>: CVideoPlayer::OpenStream - Unsupported stream 3001. Stream disabled.
2024-05-22 18:20:40.441 T:24788    info <general>: Opening stream: 3005 source: 256
2024-05-22 18:20:40.442 T:24788   debug <general>: AddOnLog: inputstream.adaptive: OpenStream(3005)
2024-05-22 18:20:40.448 T:24788   debug <general>: CurlFile::XFILE::CCurlFile::Open - <https://
2024-05-22 18:20:40.629 T:24788   debug <general>: AddOnLog: inputstream.adaptive: Download finished: https://
2024-05-22 18:20:40.631 T:24788   debug <general>: AddOnLog: inputstream.adaptive: Manifest saved to: D:\KodiPortable_Builded_20\portable_data\userdata\addon_data\inputstream.adaptive\manifests\manifest_1716394840_child-video.txt
2024-05-22 18:20:40.633 T:24788   debug <general>: AddOnLog: inputstream.adaptive: GetStream(3005)
2024-05-22 18:20:40.633 T:24788   error <general>: CVideoPlayerVideo::OpenStream: Codec id 27 require extradata.
2024-05-22 18:20:40.634 T:24788 warning <general>: CVideoPlayer::OpenStream - Unsupported stream 3005. Stream disabled.
2024-05-22 18:20:40.634 T:24788    info <general>: Opening stream: 3002 source: 256
2024-05-22 18:20:40.634 T:24788   debug <general>: AddOnLog: inputstream.adaptive: OpenStream(3002)
2024-05-22 18:20:40.639 T:24788   debug <general>: CurlFile::XFILE::CCurlFile::Open - <https://
2024-05-22 18:20:40.989 T:24788   debug <general>: AddOnLog: inputstream.adaptive: Download finished: https://
2024-05-22 18:20:40.995 T:24788   debug <general>: AddOnLog: inputstream.adaptive: Manifest saved to: D:\KodiPortable_Builded_20\portable_data\userdata\addon_data\inputstream.adaptive\manifests\manifest_1716394840_child-video_1.txt
2024-05-22 18:20:41.002 T:24788   debug <general>: AddOnLog: inputstream.adaptive: GetStream(3002)
2024-05-22 18:20:41.003 T:24788   error <general>: CVideoPlayerVideo::OpenStream: Codec id 27 require extradata.
2024-05-22 18:20:41.003 T:24788 warning <general>: CVideoPlayer::OpenStream - Unsupported stream 3002. Stream disabled.
2024-05-22 18:20:41.004 T:24788    info <general>: Opening stream: 3003 source: 256
2024-05-22 18:20:41.004 T:24788   debug <general>: AddOnLog: inputstream.adaptive: OpenStream(3003)
2024-05-22 18:20:41.016 T:24788   debug <general>: CurlFile::XFILE::CCurlFile::Open - <https://
2024-05-22 18:20:41.385 T:24788   debug <general>: AddOnLog: inputstream.adaptive: Download finished: https://
2024-05-22 18:20:41.387 T:24788   debug <general>: AddOnLog: inputstream.adaptive: Manifest saved to: D:\KodiPortable_Builded_20\portable_data\userdata\addon_data\inputstream.adaptive\manifests\manifest_1716394841_child-video.txt
2024-05-22 18:20:41.391 T:24788   debug <general>: AddOnLog: inputstream.adaptive: GetStream(3003)
2024-05-22 18:20:41.391 T:24788   error <general>: CVideoPlayerVideo::OpenStream: Codec id 27 require extradata.
2024-05-22 18:20:41.391 T:24788 warning <general>: CVideoPlayer::OpenStream - Unsupported stream 3003. Stream disabled.
2024-05-22 18:20:41.391 T:24788    info <general>: Opening stream: 3004 source: 256
2024-05-22 18:20:41.391 T:24788   debug <general>: AddOnLog: inputstream.adaptive: OpenStream(3004)
2024-05-22 18:20:41.395 T:24788   debug <general>: CurlFile::XFILE::CCurlFile::Open - <https://
2024-05-22 18:20:41.747 T:24788   debug <general>: AddOnLog: inputstream.adaptive: Download finished: https://
2024-05-22 18:20:41.749 T:24788   debug <general>: AddOnLog: inputstream.adaptive: Manifest saved to: D:\KodiPortable_Builded_20\portable_data\userdata\addon_data\inputstream.adaptive\manifests\manifest_1716394841_child-video_1.txt
2024-05-22 18:20:41.751 T:24788   debug <general>: AddOnLog: inputstream.adaptive: GetStream(3004)
2024-05-22 18:20:41.752 T:24788   error <general>: CVideoPlayerVideo::OpenStream: Codec id 27 require extradata.
2024-05-22 18:20:41.752 T:24788 warning <general>: CVideoPlayer::OpenStream - Unsupported stream 3004. Stream disabled.
2024-05-22 18:20:41.752 T:24788    info <general>: Closing stream player 2
```